src/optimal_image/search.rs

```diff
@@ -1,37 +1,132 @@
+            image_data: original,
+            options: options,
+            result: None,
+        };
+
+        Ok(search)
+    }
+
+    pub fn run(&mut self) -> Option<SearchResult> {
+        // let mut config_map: HashMap<&ImageFormat, Vec<ImageConfig>> = HashMap::new();
+        let configs = self.options.formats.iter().fold(
+            HashMap::new(),
+            |mut map: HashMap<&ImageFormat, Vec<ImageConfig>>, format| {
+                let conf_list = (self.options.quality_range.0
+                    ..self.options.quality_range.1)
+                    .map(|q| {
+                        let fmt = format.clone();
+                        ImageConfig {
+                            id: self.path.clone(),
+                            encoding_config: EncodingConfig {
+                                format: fmt,
+                                quality: q,
+                            },
+                        }
+                    })
+                    .collect();
+
+                map.insert(format, conf_list);
+                map
+            },
+        );
+
+        let result: SearchResult = configs.iter().fold(
+            HashMap::new(),
+            |mut result: SearchResult, (&k, v)| {
+                let optimal_config = find_optimal_config(
+                    &self,
+                    &v,
+                    self.options.threshold,
+                );
+                result.insert(k.clone(), optimal_config.clone());
+                result
+            },
+        );
+
+        println!("{:?}", result);
+
+        self.result = Some(result.clone());
+
+        Some(result)
+    }
+
+    pub fn get_result(&mut self) -> &Option<SearchResult> {
+        &self.result
     }
 }
 
-pub fn find_optimal_config(
-    config: &Vec<ImageConfig>,
+pub fn find_optimal_config<'a>(
+    search: &Search,
+    config: &'a &Vec<ImageConfig>,
     threshold: f64,
-) -> &ImageConfig {
+) -> &'a ImageConfig {
+    let image_jpeg_image = ImageJpegImage::from_image_data(&search.image_data).unwrap();
     let index = config
         .binary_search_by(|probe| {
-            // TODO:
-            // generate image with `probe` config
-            // run dssim against original
-            // return Ordering according to threshold
-            unimplemented!();
+            let mut dssim_context = dssim::new();
+            let original = dssim_context.create_image(&search.image_data).unwrap();
+            let q = probe.encoding_config.quality;
+            let jpeg = image_jpeg_image.encode(q).unwrap().to_image_data().unwrap();
+            let variant = dssim_context.create_image(&jpeg).unwrap();
+            let (val, _) = dssim_context.compare(&original, variant);
+            println!("comparing q: {} val: {}", q, val);
+            let diff = val - threshold;
+            if diff < 0.0 {
+                return Ordering::Greater;
+            }
+            Ordering::Equal
         })
         .unwrap();
```

Comment:
this unwrap should be handled properly: if the binary search fails it means that the threshold is too high: trying to get a too shitty image, in that case should return the lower quality in the range.